Compiler/compiler.cpp

```diff
@@ -0,0 +1,244 @@
+    ccSetOption(SCOPT_OLDSTRINGS, !comp_opts.Flags.EnforceNewStrings);
+
+    ccRemoveDefaultHeaders();
+
+    //-----------------------------------------------------------------------//
+    // Read input files
+    //-----------------------------------------------------------------------//
+    std::vector<std::pair<String, String>> heads;
+    for(const auto& header: comp_opts.HeaderFiles)
+    {
+        Stream *in = File::OpenFileRead(header.c_str());
+        if (!in)
+        {
+            std::cerr << "Error: failed to open header for reading: " << header << std::endl;
+            return -1;
+        }
+
+        String headername = Path::GetFilename(header.c_str());
+        headername = Path::RemoveExtension(headername);
+
+        TextStreamReader sr(in);
+        heads.emplace_back(sr.ReadAll(), headername);
+        in->Close();
+    }
+
+    String script_input;
+    const char* src = nullptr;
+    if(!comp_opts.InputScriptFile.empty())
+    {
+        src = comp_opts.InputScriptFile.c_str();
+        Stream *in = File::OpenFileRead(src);
+        if (!in)
+        {
+            std::cerr << "Error: failed to open script for reading: " << src << std::endl;
+            return -1;
+        }
+        TextStreamReader sr(in);
+        script_input = sr.ReadAll();
+        in->Close();
+    }
+
+    //-----------------------------------------------------------------------//
+    // Preprocess headers
+    //-----------------------------------------------------------------------//
+    std::vector<std::pair<String, String>> preprocessed_heads;
+    for(const auto& head: heads)
+    {
+        String preprocessed_header = pp.Preprocess(head.first,head.second);
+        preprocessed_heads.emplace_back(preprocessed_header.GetCStr(),head.second);
+
+        ccAddDefaultHeader((char *) preprocessed_heads.back().first.GetCStr(), (char *) preprocessed_heads.back().second.GetCStr());
+    }
+    heads.clear();
+
+    //-----------------------------------------------------------------------//
+    // Preprocess script
+    //-----------------------------------------------------------------------//
+    String script_pp = nullptr;
+    String filename = Path::GetFilename(comp_opts.InputScriptFile.c_str());
+    String script_name = Path::RemoveExtension(filename);
+
+    script_pp = pp.Preprocess(script_input,script_name);
+    if ((script_pp == nullptr) || (ccError != 0))
+    {
+        std::cerr << "Error: preprocessor failed at " << script_name.GetCStr() <<
+            ", line " << ccErrorLine << " : " << ccErrorString.GetCStr() << std::endl;
+        return -1;
+    }
+
+    if(comp_opts.PreprocessOnly)
+    {
+        Stream *out = File::CreateFile(comp_opts.OutputObjFile.c_str());
+        if (!out || !(out->CanWrite())) {
+            std::cerr << "Error: failed to open for writing: " << comp_opts.OutputObjFile << std::endl;
+            return -1;
+        }
+        script_pp.Write(out);
+        out->Close();
```

Comment:
`out` will not be deleted here. In practice this is not serious, because you are exiting right away, but in any other situation this will lead to a memory leak.

In the engine we usually have `delete out`, because stream calls Close in destructor, but engine's code is old and often not using modern C++. The right thing here would be to declare `std::unique_ptr<Stream>` instead, that would deal with both closing and deletion.

Reply:
I did exactly it using unique_ptr https://github.com/adventuregamestudio/ags/pull/1370/commits/a922cee7f09b669c4b0073e75061cc14048ca579#diff-0cd44e2bb5fd3e84785ae6ad43582a662c32f9c080dae17bf0794315f2daf1eeL236-R241